velero-plugin-for-aws/object_store.go

```diff
@@ -233,6 +236,10 @@ func newSessionOptions(config aws.Config, profile string, caCert string, credent
 			return session.Options{}, errors.Wrapf(err, "could not get credentialsFile info")
 		}
 		sessionOptions.SharedConfigFiles = []string{credentialsFile}
+
+		if sharedConfig, berr := strconv.ParseBool(enableSharedConfig); sharedConfig && berr != nil {
+			sessionOptions.SharedConfigState = session.SharedConfigEnable
```

Comment:
It's not clear to me what setting this value enables. Looking at the [docs](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/sessions.html#shared-configuration-fields), it seems that when this is set, the client will also attempt to read configuration from `~/.aws/config`. This file will not be present in the Velero pod, so I think setting this won't have an effect. Velero also doesn't use the shared credentials file. In the case where a set of credentials is set for a particular BSL, we serialise those credentials to disk and use that path in `SharedConfigFiles` in the session options. My understanding is that that takes precedence over any credentials/config set by environment variables. Please let me know if I've misunderstood this or am missing something :)

Reply:
Shared config files are different than regular credentials files: 
https://docs.aws.amazon.com/sdkref/latest/guide/file-format.html

While they can have very similar values they have some extensions:

```
[profile developers]
role_arn = arn:aws:iam::123456789012:role/developers
source_profile = dev-user
region = us-west-2
output = json
```

And it allows you to request to act like a specific role using some other auth providers such as OIDC afaik.

Reply:
I understand that difference between these files and the standard credentials files, but the part I'm confused about is how a shared config file will be made available to the plugin process. Is this setting needed to enable the AWS SDK to parse and use files of this format? For example, if you were able to pass a file of this format into `sessionOptions.SharedConfigFiles`, would it fail to be used without setting `ShareConfigState`?

Reply:
>  but the part I'm confused about is how a shared config file will be made available to the plugin process.

You can pass the shared config file into the secret and use that using the `Credentials` field. 

> Is this setting needed to enable the AWS SDK to parse and use files of this format? For example, if you were able to pass a file of this format into sessionOptions.SharedConfigFiles, would it fail to be used without setting ShareConfigState?

It does fail to load unless you set it explicitly to try and read a config file. Also in my testing, if this is set, it fails to read a credentials file format. 

Sorry if I am misunderstanding the major concern and not doing a good job explaining it.

Reply:
Summarising a discussion with @shawn-hurley on Slack:

Enabling `SharedConfigState` allows the SDK to parse [Shared Config Files](https://docs.aws.amazon.com/sdkref/latest/guide/file-format.html) when they are provided through the `SharedConfigFiles` session option. This field is only set when a BSL is using its [own credentials](https://github.com/vmware-tanzu/velero-plugin-for-aws#create-an-additional-backup-storage-location), not those provided by the Velero deployment. Currently, if a shared config file is provided in this way, it will fail to load. Enabling this setting allows Shared Config Files to be used.

This setting must be enabled per BSL/VSL as it determines how the AWS SDK will parse the credential/configuration file used for that particular BSL/VSL.

This is currently only available for BSLs as they support unique credentials. We will be adding support for per-VSL credentials in vmware-tanzu/velero#4115.

@shawn-hurley: Please add any other details that I missed :)